uat/testing-features/src/main/java/com/aws/greengrass/aws/IAMRole.java

```diff
@@ -0,0 +1,181 @@
+import software.amazon.awssdk.services.iam.model.CreatePolicyResponse;
+import software.amazon.awssdk.services.iam.model.CreateRoleRequest;
+import software.amazon.awssdk.services.iam.model.CreateRoleResponse;
+import software.amazon.awssdk.services.iam.model.DeletePolicyVersionRequest;
+import software.amazon.awssdk.services.iam.model.DeleteRoleRequest;
+import software.amazon.awssdk.services.iam.model.DetachRolePolicyRequest;
+import software.amazon.awssdk.services.iam.model.GetRoleRequest;
+import software.amazon.awssdk.services.iam.model.IamException;
+import software.amazon.awssdk.services.iam.model.ListAttachedRolePoliciesRequest;
+import software.amazon.awssdk.services.iam.model.GetPolicyRequest;
+import software.amazon.awssdk.services.iam.model.GetPolicyVersionRequest;
+import software.amazon.awssdk.services.iam.model.CreatePolicyVersionResponse;
+import software.amazon.awssdk.services.iam.model.CreatePolicyVersionRequest;
+
+import static com.aws.greengrass.aws.IotShadowResources.INITIAL_RETRY_INTERVAL_SEC;
+import static com.aws.greengrass.aws.IotShadowResources.MAX_RETRY_INTERVAL_SEC;
+import static com.aws.greengrass.aws.IotShadowResources.MAX_RETRY_NUM;
+import static com.aws.greengrass.aws.IotShadowResources.getIamTag;
+import static com.aws.greengrass.aws.IotShadowResources.randomName;
+
+@Data
+@Log4j2
+public class IAMRole implements IotShadowResource<IamClient> {
+
+    String iamRoleArn;
+    String iamRoleName;
+    boolean deleted = false;
+    boolean persist;
+    static final ObjectMapper mapper = new ObjectMapper();
+
+
+    IAMRole(CreateRoleResponse response, boolean persist) {
+        this.iamRoleName = response.role().roleName();
+        this.iamRoleArn = response.role().arn();
+        this.persist = persist;
+    }
+
+    static Result<Pair<IAMRole, IAMPolicy>> create(IamClient client, IAMRoleSpec spec) throws InterruptedException {
+        if (spec.getRoleName() == null) {
+            spec.setRoleName(randomName());
+        }
+        IAMRole role = null;
+        String desiredPolicyDocument = spec.getPolicyDocument();
+
+        // try get role first.
+        try {
+            role = new IAMRole(CreateRoleResponse.builder()
+                    .role(client.getRole(GetRoleRequest.builder().roleName(spec.getRoleName()).build()).role()).build(),
+                    spec.isPersist());
+            // if role exists, check and update the policy if needed
+            if (role != null) {
+                // check if desired policy is attached
+                for (AttachedPolicy p : client.listAttachedRolePoliciesPaginator(
+                        ListAttachedRolePoliciesRequest.builder().roleName(spec.getRoleName()).build()).attachedPolicies()) {
+                    if (p.policyName().equals(spec.getPolicyName())) {
+                        //policy attached
+                        String defaultVersionId = client.getPolicy(GetPolicyRequest.builder()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code appears to be waiting for a resource before it runs. You could use the waiters feature to help improve efficiency. Consider using PolicyExists. For more information, see [https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/](https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/)